File: dsa/linked_list/206_Reverse_Linked_List.java
package linked_list;

//  * Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null){
            return null;
        }

        if (head.next == null){
            return head;
        }
        ListNode prevNode = null;
        ListNode indNode = head;
        while (indNode != null){
            ListNode nextNode = indNode.next;
            indNode.next = prevNode;
            prevNode = indNode;
            indNode = nextNode;
        }
        return prevNode;
    }
}

File: dsa/7+8.linked_list/206_Reverse_Linked_List.py
import typing as t
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseList(self, head: t.Optional[ListNode]) -> t.Optional[ListNode]:
        prev_node: ListNode = None
        pointer_node: ListNode = head

        while pointer_node:
            next_node: ListNode = pointer_node.next
            pointer_node.next = prev_node
            prev_node = pointer_node
            pointer_node = next_node
            
        return prev_node
        